### PR TITLE
adds GDK suspend/resume basic handling

### PR DESCRIPTION
**Pull request by cigumo**
Adds support for GDK suspend/resume for Xbox. 

I don't necessarily mean that this should be merged into SDL2. I wanted to leave it on the record here, in case anyone needs to use it in their GDK Xbox games.

The application will receive the suspend/resume as SDL events (SDL_APP_DIDENTERBACKGROUND, SDL_APP_WILLENTERFOREGROUND).
When the suspend event is received, the application must call SDL_GDKSuspendComplete() within one second (after calling D3D SuspendX() for example) to let the suspend process continue. 

Having a generic way for the applications to notify SDL when the resume process is finished from their side is something that would be good to have in SDL3.

### CODE REVIEW

File: src/core/gdk/SDL_gdk.cpp [hunk 1]
```diff
@@ -20,16 +20,24 @@
...
+}
 #include <XGameRuntime.h>
 #include <xsapi-c/services_c.h>
 #include <shellapi.h> /* CommandLineToArgvW() */
+#include <appnotify.h>
 
 static XTaskQueueHandle GDK_GlobalTaskQueue;
 
+PAPPSTATE_REGISTRATION hPLM = {};
+HANDLE plmSuspendComplete = nullptr;
```
Comment (flibitijibibo) [inline src/core/gdk/SDL_gdk.cpp:39]: Is it possible for these to be static?

File: src/core/gdk/SDL_gdk.cpp [hunk 2]
```diff
@@ -144,9 +152,40 @@ SDL_GDKRunApp(SDL_main_func mainFunction, void *reserved)
...
+        }
+        auto rascn = [](BOOLEAN quiesced, PVOID context)
+        {
+            SDL_LogDebug(SDL_LOG_CATEGORY_APPLICATION, "[GDK] in RegisterAppStateChangeNotification handler");
+            if (quiesced) {
+                ResetEvent(plmSuspendComplete);
+                SDL_SendAppEvent(SDL_APP_DIDENTERBACKGROUND);
+
+                // To defer suspension, we must wait to exit this callback.
+                // IMPORTANT: The app must call SDL_GDKSuspendComplete() to release this lock.
```
Comment (flibitijibibo) [inline src/core/gdk/SDL_gdk.cpp:169]: For maintainers: Do we care if C++ files use C++ comments?

**Review by flibitijibibo — APPROVED** [review]
A couple nits (half of it is for the maintainers, honestly), otherwise makes sense to me. GDK should be able to handle the C++-isms in the patch just fine!

File: src/core/gdk/SDL_gdk.cpp [hunk 3]
```diff
@@ -20,16 +20,24 @@
 */
 #include "../../SDL_internal.h"
 
+extern "C" {
```
Comment (sezero) [inline src/core/gdk/SDL_gdk.cpp:23]: Why did we really not have this previously? (And how did it work?)

File: src/dynapi/SDL_dynapi_overrides.h [hunk 4]
```diff
@@ -877,6 +877,7 @@
 #define SDL_utf8strnlen SDL_utf8strnlen_REAL
 #define SDL_GDKGetTaskQueue SDL_GDKGetTaskQueue_REAL
 #define SDL_GDKRunApp SDL_GDKRunApp_REAL
+#define SDL_GDKSuspendComplete SDL_GDKSuspendComplete_REAL
```
Comment (sezero) [inline src/dynapi/SDL_dynapi_overrides.h:880]: Do not insert: Append to end.

File: src/dynapi/SDL_dynapi_procs.h [hunk 5]
```diff
@@ -959,6 +959,7 @@ SDL_DYNAPI_PROC(size_t,SDL_utf8strnlen,(const char *a, size_t b),(a,b),return)
 #if defined(__GDK__)
 SDL_DYNAPI_PROC(int,SDL_GDKGetTaskQueue,(XTaskQueueHandle *a),(a),return)
 SDL_DYNAPI_PROC(int,SDL_GDKRunApp,(SDL_main_func a, void *b),(a,b),return)
+SDL_DYNAPI_PROC(void,SDL_GDKSuspendComplete,(void),(),return)
```
Comment (sezero) [inline src/dynapi/SDL_dynapi_procs.h:962]: Same as above

File: src/dynapi/SDL2.exports [hunk 6]
```diff
@@ -851,6 +851,7 @@
 ++'_SDL_utf8strnlen'.'SDL2.dll'.'SDL_utf8strnlen'
 # ++'_SDL_GDKGetTaskQueue'.'SDL2.dll'.'SDL_GDKGetTaskQueue'
 # ++'_SDL_GDKRunApp'.'SDL2.dll'.'SDL_GDKRunApp'
+# ++'_SDL_GDKSuspendComplete'.'SDL2.dll'.'SDL_GDKSuspendComplete'
```
Comment (sezero) [inline src/dynapi/SDL2.exports:854]: Same

**Comment by sezero** [comment]
Pushed https://github.com/libsdl-org/SDL/commit/cd399caf6f4921b59fd8ba08c583ec8df70ef003 as a follow-up to fix dynapi entries.

**Comment by sezero** [comment]
Will this go into SDL3 or will there be a generic way of handling the issue?

File: src/joystick/SDL_gamecontrollerdb.h [hunk 7]
```diff
@@ -654,8 +654,6 @@ static const char *s_ControllerMappings [] =
...
     "030000006d04000019c2000011010000,Logitech F710 Gamepad (DInput),a:b1,b:b2,back:b8,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,leftshoulder:b4,leftstick:b10,lefttrigger:b6,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b11,righttrigger:b7,rightx:a2,righty:a3,start:b9,x:b0,y:b3,", /* Guide button doesn't seem to be sent in DInput mode. */
     "030000006d0400001fc2000005030000,Logitech F710 Gamepad (XInput),a:b0,b:b1,back:b6,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b8,leftshoulder:b4,leftstick:b9,lefttrigger:a2,leftx:a0,lefty:a1,rightshoulder:b5,rightstick:b10,righttrigger:a5,rightx:a3,righty:a4,start:b7,x:b2,y:b3,",
-    "030000006d0400004fc2000011010000,Logitech G29 Racing Wheel,a:b0,b:b2,back:b8,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b24,leftshoulder:b5,leftstick:b11,lefttrigger:b7,rightshoulder:b4,rightstick:b10,righttrigger:b6,start:b9,x:b1,y:b3,",
-    "030000006d04000060c2000010010000,Logitech G29 Racing Wheel,a:b1,b:b2,back:b8,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,dpup:h0.1,guide:b12,leftshoulder:b4,leftstick:b10,lefttrigger:b6,rightshoulder:b5,rightstick:b11,righttrigger:b7,start:b9,x:b0,y:b3,",
```
Comment (tomcashman) [inline src/joystick/SDL_gamecontrollerdb.h:658]: Were these meant to be removed?

File: src/joystick/controller_type.c [hunk 8]
```diff
@@ -32,7 +32,7 @@ static const ControllerDescription_t arrControllers[] = {
 	{ MAKE_CONTROLLER_ID( 0x0079, 0x1844 ), k_eControllerType_PS3Controller, NULL },	// From SDL
 	{ MAKE_CONTROLLER_ID( 0x044f, 0xb315 ), k_eControllerType_PS3Controller, NULL },	// Firestorm Dual Analog 3
 	{ MAKE_CONTROLLER_ID( 0x044f, 0xd007 ), k_eControllerType_PS3Controller, NULL },	// Thrustmaster wireless 3-1
-	{ MAKE_CONTROLLER_ID( 0x046d, 0xc24f ), k_eControllerType_PS3Controller, NULL },	// Logitech G29 (PS3)
+	//{ MAKE_CONTROLLER_ID( 0x046d, 0xc24f ), k_eControllerType_PS3Controller, NULL },	// Logitech G29 (PS3)
```
Comment (tomcashman) [inline src/joystick/controller_type.c:35]: Was this meant to be removed?

File: src/joystick/controller_type.c [hunk 9]
```diff
@@ -87,7 +87,7 @@ static const ControllerDescription_t arrControllers[] = {
 	{ MAKE_CONTROLLER_ID( 0x8888, 0x0308 ), k_eControllerType_PS3Controller, NULL },	// Sony PS3 Controller
 
 	{ MAKE_CONTROLLER_ID( 0x0079, 0x181b ), k_eControllerType_PS4Controller, NULL },	// Venom Arcade Stick - XXX:this may not work and may need to be called a ps3 controller
-	{ MAKE_CONTROLLER_ID( 0x046d, 0xc260 ), k_eControllerType_PS4Controller, NULL },	// Logitech G29 (PS4)
+	//{ MAKE_CONTROLLER_ID( 0x046d, 0xc260 ), k_eControllerType_PS4Controller, NULL },	// Logitech G29 (PS4)
```
Comment (tomcashman) [inline src/joystick/controller_type.c:90]: Same as above.

**Pull request merged**